src/formatters/mdapi/retrieveResultFormatter.ts

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { join, parse } from 'path';
+import { blue, yellow } from 'chalk';
+import { UX } from '@salesforce/command';
+import { Logger, Messages, SfdxError } from '@salesforce/core';
+import { get, getNumber } from '@salesforce/ts-types';
+import {
+  RetrieveResult,
+  MetadataApiRetrieveStatus,
+  RequestStatus,
+  RetrieveMessage,
+  FileProperties,
+} from '@salesforce/source-deploy-retrieve';
+import { ResultFormatter, ResultFormatterOptions, toArray } from '../resultFormatter';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/plugin-source', 'retrieve');
+
+export type RetrieveCommandResult = Omit<MetadataApiRetrieveStatus, 'zipFile'> & { zipFilePath: string };
+
+export interface RetrieveResultFormatterOptions extends ResultFormatterOptions {
+  retrieveTargetDir: string;
+  zipFileName?: string;
+  unzip?: boolean;
+}
+
+export interface RetrieveCommandAsyncResult {
+  done: boolean;
+  id: string;
+  state: RequestStatus | 'Queued';
+  status: RequestStatus | 'Queued';
+  timedOut: boolean;
+}
+
+export class RetrieveResultFormatter extends ResultFormatter {
+  public options: RetrieveResultFormatterOptions;
+  protected result: RetrieveCommandResult;
+  protected warnings: RetrieveMessage[];
+
+  public constructor(logger: Logger, ux: UX, options: RetrieveResultFormatterOptions, result: RetrieveResult) {
+    super(logger, ux, options);
+    // zipFile can become massive and unwieldy with JSON parsing/terminal output and, isn't useful
+    delete result.response.zipFile;
+    this.options.zipFileName ??= 'unpackaged.zip';
+    const zipFilePath = join(options.retrieveTargetDir, options.zipFileName);
+    this.result = Object.assign({}, result.response, { zipFilePath });
+
+    // grab warnings
+    const warnMessages = get(result, 'response.messages', []) as RetrieveMessage | RetrieveMessage[];
```

Comment:
are the SDR types wrong that we have to use brittle `get` and casting?

could this not just say
```ts
this.warnings = toArray(result.response.messages ?? [])
```

Reply:
That line is copied from source:retrieve.  Yes, I'd love to rely on SDR types more but they're not always friendly to work with.  I can update.